api/internal/localizer/locloader.go

```diff
@@ -0,0 +1,382 @@
+		if ll.local && dir.HasPrefix(ll.args.newDir) {
+			return nil, nil, errors.WrapPrefixf(errDstRef, errReference, cleanPath)
+		}
+		return content, &LocPath{false, pathSuffix(ll.ldr.Root(), cleanPath)}, nil
+	}
+
+	return nil, nil, errors.WrapPrefixf(errors.Errorf(errAlpha), "path '%s' is absolute", path)
+}
+
+func parseDomain(host string) string {
+	if host == "gh:" {
+		return "github.com"
+	}
+	target := host
+	// remove scheme from target
+	for _, p := range git.Schemes() {
+		if strings.HasPrefix(target, p) {
+			target = target[len(p):]
+			break
+		}
+	}
+	// remove userinfo
+	if i := strings.Index(target, "@"); i > -1 {
+		target = target[i+1:]
+	}
+	// remove ssh path delimiter or port delimiter
+	if i := strings.Index(target, ":"); i > -1 {
+		target = target[:i]
+	}
+	// remove http path delimiter
+	return strings.TrimSuffix(target, "/")
+}
+
+// handleNewLocalDir returns the relative path between the current LocLoader root and the new root at ldr
+// if the new root satisfies localize requirements. Otherwise, handleNewLocalDir returns an error.
+func (ll *LocLoader) handleNewLocalDir(ldr ifc.Loader) (string, error) {
+	if ll.local && !filesys.ConfirmedDir(ldr.Root()).HasPrefix(ll.args.scope) {
+		return "", errors.WrapPrefixf(
+			errRootOutsideScope,
+			"invalid root location '%s' for scope '%s'", ldr.Root(), ll.args.scope)
+	}
+	if ll.local && filesys.ConfirmedDir(ldr.Root()).HasPrefix(ll.args.newDir) {
+		return "", errors.WrapPrefixf(errDstRef, errReference, ldr.Root())
+	}
+	rel, err := filepath.Rel(ll.ldr.Root(), ldr.Root())
+	if err != nil {
+		log.Fatalf(errors.WrapPrefixf(err,
+			"%s: cannot find relative path from root '%s' to its valid root reference '%s'",
+			fatalPrefix, ll.ldr.Root(), ldr.Root()).Error())
+	}
+	return rel, nil
+}
+
+// New returns a LocLoader at path and whether path is remote if path is a root that can be localized.
+// Otherwise, New returns an error.
+func (ll *LocLoader) New(path string) (*LocLoader, *LocPath, error) {
+	ldr, err := ll.ldr.New(path)
+	if err != nil {
+		return nil, nil, errors.WrapPrefixf(err, "invalid root reference")
+	}
+	var repo, inDst string
+	var isRemote bool
+	if repo, isRemote = ldr.Repo(); isRemote {
+		spec, err := parseLocRootURL(path)
+		if err != nil {
+			return nil, nil, err
+		}
+		inRepo := pathSuffix(repo, ldr.Root())
+		inDst = filepath.Join(localizeDir, parseDomain(spec.Host), toFSPath(spec.OrgRepo), toFSPath(spec.Ref), inRepo)
+	} else {
+		inDst, err = ll.handleNewLocalDir(ldr)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+	return &LocLoader{
+			fSys:  ll.fSys,
+			args:  ll.args,
+			ldr:   ldr,
+			local: ll.local && !isRemote,
+			dst:   filepath.Join(ll.dst, inDst),
+		}, &LocPath{
+			isRemote,
+			inDst,
+		}, nil
+}
+
+// Cleanup tries to clean the by-products of localize and logs if unsuccessful
+func (ll *LocLoader) Cleanup() {
+	cleanLdr(ll.ldr)
+}
+
+func cleanLdr(ldr ifc.Loader) {
+	if err := ldr.Cleanup(); err != nil {
```

Comment:
The way that this is written is pretty convoluted. I suggest we just use `ldr.Cleanup` wherever you are currently using `cleanLdr`. I understand you are concerned about potentially returning two errors. This is one special case that I think it is appropriate and best to ignore the return value of `ldr.Cleanup`. 

Btw, if we embed the loader as I suggested above, we will not need to define an additional `LocLoader.Cleanup`.

Reply:
Like you suggested, I removed `LocLoader.Cleanup()` so that the calling code will use `ldr.Cleanup()` instead. I wasn't sure if you meant that logging the `Cleanup()` error in the calling code is still "convoluted", so I kept it in `ValidateLocArgs`. 

If you meant that we shouldn't log the `Cleanup()` error in general, I can remove the corresponding lines in `ValidateLocArgs`.

Reply:
Per request, I've removed the logging for `ldr.Cleanup()` in the latest commit.